File: TargetSum3.cpp
#include<iostream>
#include<vector>
#include<unordered_map>

using namespace std;

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int S) {
    	int n = nums.size();
    	unordered_map<int, int> dp;
    	dp[0] = 1;
    	for(int i = 0; i < n; i++) {
    		unordered_map<int, int> t;
    		for(auto &a : dp) {
    			int sum = a.first, count = a.second;
    			t[sum - nums[i]] += count;
    			t[sum + nums[i]] += count;
			}
			dp = t;
		}
		return dp[S];
    }
};

int main() {
	
	int n;
	cin>>n;
	
	vector<int> nums;
	for(int i = 0; i < n; i++) {
		int num;
		cin>>num;
		nums.push_back(num);
	}
	
	int S;
	cin>>S;
	
	Solution *solution = new Solution();
	cout<<solution->findTargetSumWays(nums, S);
	
	return 0;
}

File: TargetSum3.py
class Solution:
    def findTargetSumWays(self, nums: list, S: int) -> int:
        s = sum(nums)
        if s < S:
            return 0
        dp = [0] * (2 * s + 1)
        dp[s] = 1
        for num in nums:
            t = [0] * (2 * s + 1)
            for i in range(num, 2 * s + 1 - num):
                t[i - num] += dp[i]
                t[i + num] += dp[i]
            dp = t
        return dp[S + s]


if __name__ == '__main__':
    nums = [int(num) for num in input().split()]
    S = int(input())
    solution = Solution()
    print(solution.findTargetSumWays(nums, S))
